integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -77,88 +292,143 @@ private class ListenableFutureCoroutine<T>(
 }
 
 /**
- * Converts this deferred value to the instance of [ListenableFuture].
- * The deferred value is cancelled when the resulting future is cancelled or otherwise completed.
+ * A [ListenableFuture] that delegates to an internal [DeferredListenableFuture], collaborating with
+ * it.
+ *
+ * This setup allows the returned [ListenableFuture] to maintain the following properties:
+ *
+ * - Correct implementation of [Future]'s happens-after semantics documented for [get], [isDone]
+ *   and [isCancelled] methods
+ * - Cancellation propagation both to and from [Deferred]
+ * - Correct cancellation and completion semantics even when this [ListenableFuture] is combined
+ *   with different concrete implementations of [ListenableFuture]
+ *   - Fully correct cancellation and listener happens-after obeying [Future] and
+ *     [ListenableFuture]'s documented and implicit contracts is surprisingly difficult to achieve.
+ *     The best way to be correct, especially given the fun corner cases from
+ *     [AsyncFuture.setAsync], is to just use an [AsyncFuture].
+ *   - To maintain sanity, this class implements [ListenableFuture] and uses an inner [AsyncFuture]
+ *     around its input [deferred] as a state engine to establish happens-after-completion. This
+ *     could probably be compressed into one subclass of [AsyncFuture] to save an allocation, at the
+ *     cost of the implementation's readability.
  */
-public fun <T> Deferred<T>.asListenableFuture(): ListenableFuture<T> = DeferredListenableFuture<T>(this)
+private class OuterFuture<T>(val deferred: Deferred<T>): ListenableFuture<T> {
```

Comment:
`private val deferred` will save two methods in the bytecode :)

Reply:
Is there a YouTrack for fixing this to happen automatically? Even if it's in the new IR backend only, you shouldn't be required to know things like this.

Reply:
Done.

Reply:
@JakeWharton See https://youtrack.jetbrains.com/issue/KT-16454